src/include/mpi.h.in

```diff
@@ -806,6 +811,11 @@ typedef int (MPI_Delete_function) ( MPI_Comm, int, void *, void * );
 /* User combination function */
 typedef void (MPI_User_function) ( void *, void *, int *, MPI_Datatype * );
 typedef void (MPI_User_function_c) ( void *, void *, MPI_Count *, MPI_Datatype * );
+/* the _x callbacks accepts scalar inputs and an additional "void *" extra_state */
+typedef void (MPIX_User_function_x) ( void *, void *, MPI_Count, MPI_Datatype, void *);
```

Comment:
Any chance we could call the typedef `MPIX_Reduce_function` ? I always found the bare `User` somewhat of a misnomer, it is a "user" thing, yes, but for what purpose? Moreover, with the new name, there is no need for the `_x` suffix.

Reply:
BTW, what about the return? I think this should return `int` to signal MPI about any errors in the execution of the callback. Errors should not pass silently, nor they should be forced to be treated as fatal.

Reply:
I find the `user` part weird as well, but I thought it should be called `MPI_Op_function`.  However, the renaming may cause alarms and we may have to add a rationale or something. Thus I think it's wise to avoid renaming ("the `User` part at lease).

What kind of errors are you having in mind? With Python or C++, you always can use exceptions, right?

Reply:
I am also in favor of `MPI_Op_function`. The type is used to describe the API and is unlikely to be used by the user. We have more than one type of "user-function" these days so we should be more descriptive. We could as well deprecate `MPI_User_function` and replace it with `MPI_Op_function` with a similar signature without breaking anything.

Reply:
> but I thought it should be called MPI_Op_function

`MPI_Op_function` works with me, too... it's just the bare "User" thing that I do not like,

> What kind of errors are you having in mind? With Python or C++, you always can use exceptions, right?

We are about to enter the endless battle where you assume without saying it that I'm a language fanatic pushing for a secret Python agenda :rofl:. 

My rationale is the following. We cannot anticipate what the user will do or must do within the callback. Executing the callback may lead to some erroneous runtime condition. If you apply the principle that a) errors should not pass silently, and b) errors should be allowed to be treated in a  non fatal way, then it follows that the user must be allow for the callback to return with error, and that error should be propagated back to the caller of the outer `MPI_Reduce` or similar call.

To put a concrete example: Imagine that to execute the reduction operation you need to offload some computation to some accelerator, and in all that mess a runtime error occurs. If the callback is not allowed to return error, then a responsible developer only have to options: a) `abort()` and everything goes to hell, or b) print to stderr and keep going (and that error may actually go unnoticed, lost in some error log file). However, if the callback is allowed to return int, then you can signal the error (maybe with some custom registered MPI error code and string) and the MPI layer propagates the error up. As per the MPI user choice, if using the fatal error handler, then the things blow, but if using the return error handler, the MPI user may be able to treat the error in a non-fatal way. In the case of MPI, that would mean converting the error to an exception, and then the Python user does her `try/except` thing to deal with it.

Reply:
> With Python or C++, you always can use exceptions, right?

I'm not sure what you mean. C++ exception handling across language boundaries is tricky. Python exception handling across language boundaries is also something I prefer to stay away. The most sane thing is for the MPI engine to propagate errors. I still cannot understand your resistance to such requirement. Maybe it just that you find it really annoying having to deal with error propagation and you would prefer to just `abort()`.  

Do you accept or not that the spirit of MPI is to allow for non fatal errors, and for proper error propagation, even when these errors originate in user callbacks? If not, why would MPI a) have errors return error handler, and b) have most callbacks return `int`, with the notable exception of the reduction callback?

Reply:
> > With Python or C++, you always can use exceptions, right?
> 
> I'm not sure what you mean.

I mean user can raise exceptions in their user function and then they can catch the exception within their language or application layer. Right?

> Maybe it just that you find it really annoying having to deal with error propagation and you would prefer to just `abort()`.

We have started that on the wrong foot, have we? 🤣 No, annoyance or implementation complexity are not behind my thinking (yet). What are behind my thinking are:
* We are changing something MPI has established for decades, what is broken with the old? -- I know you are explaining it, but I am not sure I understand or am onboard yet.
* How we, as an implementation trying to be high quality, are going to do with the user error return in this case? The reduction may happen in parallel or collectively, and I know this is a complex problem for us to tackle and it is likely to have performance consequences.
* If the best thing we can do is to throw exceptions, then the user (and the binding implementers, e.g. you) can do that already and easily. If you are looking for something better, I am not sure we can do any better, especially from an application perspective. For my understanding, I would like to examine a specific case, which leads to be question for you. It was sincere, despite your misperception.

Reply:
>  I mean user can raise exceptions in their user function and then they can catch the exception within their language or application layer. Right?

It is not guaranteed that exceptions can traverse C code (since C is not a language that supports exceptions) and it's not a mechanism that applications can rely on. Proper error propagation between op functions and would be useful (say, to catch overflows). We could introduce an error class `MPI_ERR_USER_OP` that can be returned.

I agree that ops can executed concurrently but it's not impossible to specify the behavior without much overhead: one way would be that the error returned from an invocation is cached until the operation is complete and returned as the error of that operation. Obviously, we cannot stop the operation when an operator returns an error but we don't have to. What matters is that an error that occurred is propagated.

Reply:
> It is not guaranteed that exceptions can traverse C code

I see.

Reply:
Similar concerns apply to Python, and I guess it would be a similar situation for other languages with exception-based error handling. IMHO, the sane and robust way to do thing is to use a C callback implementing a trampoline that calls into another language, and somehow you do a try/catch for exceptions, then you convert the exception to a nonzero return error code, and the MPI layer propagates the error up the chain.